server/src/main/java/keywhiz/service/resources/automation/v2/BatchResource.java

```diff
@@ -120,27 +130,38 @@ public BatchCreateOrUpdateSecretsResponseV2 batchCreateOrUpdateSecrets(
   private void createOrUpdateSecret(
       DSLContext dslContext,
       AutomationClient automationClient,
-      CreateOrUpdateSecretInfoV2 secret) {
+      CreateOrUpdateSecretInfoV2 secret,
+      SecretController.SecretBuilder builder) {
 
-    Optional<SecretSeriesAndContent> maybeSecretSeriesAndContent = secretDAO.getSecretByName(dslContext, secret.name());
     String secretOwner = secret.owner();
-    if (maybeSecretSeriesAndContent.isPresent()) {
-      permissionCheck.checkAllowedOrThrow(automationClient, Action.UPDATE, maybeSecretSeriesAndContent.get());
-    } else {
-      permissionCheck.checkAllowedForTargetTypeOrThrow(automationClient, Action.CREATE, Secret.class);
-      secretOwner = getSecretOwnerForSecretCreation(dslContext, secretOwner, automationClient);
+    {
+      Optional<SecretSeriesAndContent> maybeSecretSeriesAndContent =
+          secretDAO.getSecretByName(dslContext, secret.name());
+      if (maybeSecretSeriesAndContent.isPresent()) {
+        permissionCheck.checkAllowedOrThrow(automationClient, Action.UPDATE,
+            maybeSecretSeriesAndContent.get());
+      } else {
+        permissionCheck.checkAllowedForTargetTypeOrThrow(automationClient, Action.CREATE,
+            Secret.class);
+        secretOwner = getSecretOwnerForSecretCreation(dslContext, secretOwner, automationClient);
+      }
     }
 
-    SecretController.SecretBuilder builder = secretController
+    builder
+        .withOwnerName(secretOwner)
+        .createOrUpdate(dslContext);
+
+    emitAuditLogEntry(automationClient, secret);
+  }
+
+  private SecretController.SecretBuilder toBuilder(
+      AutomationClient automationClient,
+      CreateOrUpdateSecretInfoV2 secret) {
+    return secretController
         .builder(secret.name(), secret.content(), automationClient.getName(), secret.expiry())
         .withDescription(secret.description())
         .withMetadata(secret.metadata())
-        .withType(secret.type())
-        .withOwnerName(secretOwner);
-
-    builder.createOrUpdate(dslContext);
-
-    emitAuditLogEntry(automationClient, secret);
+        .withType(secret.type());
   }
```

Comment:
This diff is not great, but allow me to provide a little more context:

For implementations which are using a remote system (e.g. non-local HSM) for their cryptography, the initial secretController.builder() invocation on line 135/161 will invoke that system in order to compute a hash of the secret contents. The original implementation of the batch secret update method was intended to provide an atomic API for updating multiple secret updates, however, it did not take into account the possibility of making a remote call inside of a transaction. This changes hoists this potentially remote-call-invoking code outside of the transaction, which should improve throughput and availability.